uaclient/tests/test_cli_enable.py

```diff
@@ -327,12 +342,71 @@ def test_entitlements_not_found_and_beta(
                 mock.call(cfg, assume_yes=assume_yes, allow_beta=beta_flag)
             ] == m_ent_cls.call_args_list
 
-        expected_enable_call = mock.call(silent_if_inapplicable=False)
+        expected_enable_call = mock.call()
         for m_ent in mock_obj_list:
             assert [expected_enable_call] == m_ent.enable.call_args_list
 
         assert 0 == m_ent1_obj.call_count
 
+    @mock.patch("uaclient.contract.get_available_resources", return_value={})
+    @mock.patch(
+        "uaclient.entitlements.is_config_value_true", return_value=False
+    )
+    def test_entitlement_beta_and_already_enabled(
```

Comment:
@lucasmoura I added this test. Let me know if this covers what you were looking for.

Also I went with creating a custom entitlement class for this test because it seemed easier than getting the right combination of mocks, but let me know if there's a nicer way to do this with mocks. Specifically, we need:
1. the real `can_enable` (so that it can appropriately print the already enabled message)
2. `enable` to call `can_enable` and return False
3. `is_beta` to be True

Reply:
I think I have an idea to better test this, but let me know if it is an overkill here

We can split the enable operation in two steps:

1. Check if enable is possible
2. Perform enable operation

Looking at `RepoEntitlement` and `LivepatchEntitlement`, the first step is the same, they basically call the `can_enable` method. Although `RepoEntitlement` has an extra logic to ouput pre-enable messages, but this is an generic logic that could be moved into `UAEntitlement`.

Because of that, I think we can update the enable function on `UAEntitlement` to:

```python
    def enable(self, *, silent_if_inapplicable: bool = False) -> bool:
        """Enable specific entitlement.

        :param silent_if_inapplicable:
            Don't emit any messages until after it has been determined that
            this entitlement is applicable to the current machine.

        @return: True on success, False otherwise.
        @raises: UserFacingError on failure to install suggested packages
        """
        msg_ops = self.messaging.get("pre_enable", []) 
        if not handle_message_operations(msg_ops):
            return False
        if not self.can_enable(silent=silent_if_inapplicable):
            return False
        
         return self._perform_enable()

    def _perform_enable() -> bool:
        raise NotImplementedError
```

We would need to update the enable method in `RepoEntitlement` and `LivepatchEntitlement`, but the update should not be problematic. Also, If we do that, we would be able to better test this code in the unittest, since we could then just create a dummy test class like this:

```python
class TestEntitlement(entitlements.base.UAEntitlement):
    pass
```

And then, just use `mock.patch.object`  to mock the other methods as needed. But this will at least ensure that we are not producing an artificial flow for the enable operation.

But again, let me know if you think this is an overkill

Reply:
@lucasmoura I had also noticed that logic was duplicated and likely applicable to any entitlement but wasn't sure if a refactor was necessary. I like the structure you suggest here though! I'll take a pass at this in this PR - thanks!